api/routes.go

```diff
@@ -89,11 +92,30 @@ func (api *API) noAuth(h HandlerWithUser) httprouter.Handle {
 }
 
 // withAuth ensures that the user making the request has logged in.
-func (api *API) withAuth(h HandlerWithUser) httprouter.Handle {
+func (api *API) withAuth(h HandlerWithUser, allowsAPIKey bool) httprouter.Handle {
 	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
 		api.logRequest(req)
+
+		// Check for a token.
+		u, token, err := api.userAndTokenByRequestToken(req)
+		if err == nil {
+			// Embed the verified token in the context of the request.
+			ctx := jwt.ContextWithToken(req.Context(), token)
+			h(u, w, req.WithContext(ctx), ps)
+			return
+		}
+
 		// Check for an API key.
-		u, token, err := api.userAndTokenByAPIKey(req)
+		ak, err := apiKeyFromRequest(req)
+		if err != nil {
+			api.WriteError(w, err, http.StatusUnauthorized)
+			return
+		}
+		if !allowsAPIKey {
```

Comment:
move up this check? if it's not allowed we shouldn't try and parse it either

Reply:
Chris wanted a meaningful error message when a client tries to authenticate with an API key on an endpoint that doesn't support it. That's why I check for a valid API key first and then issue this special error, instead of 
```go
if !allowAPIKey {
   // respond with a 401
   return
}
```